### DISCUSSION TITLE
Transitioning Rules/Alerts to the `objstore` client

### DISCUSSION THREAD

**Discussion by jtlisi** [cortexproject/cortex#3771] [Ideas]
As discussed in past community meetings there are two Object Storage client implementations used by the Cortex project. The [chunks package implementation](https://github.com/cortexproject/cortex/blob/0976147451ee0435364b416874c3d01a0420e74f/pkg/chunk/storage_client.go#L61-L90) and the [Thanos objstore implementation](https://github.com/thanos-io/thanos/blob/8967a8cf152652ac310a8794350e9961d9e7445c/pkg/objstore/objstore.go#L34-L87).

Currently, we use the Chunks client for:
- Chunks Backend
- Rules Storage
- Alertmanager Storage
- Purger Storage

We use the blocks client for the TSDB blocks backend.

I want to take up the work of migrating the Alertmanager and Rules Storage backends to use the same client as the TSDB blocks backend. The goal of this is to:

- Make the object storage clients in the configuration file
- Reduce the surface area for required changes to a client to support different features of a backend.
- Simplify instrumentation (https://github.com/cortexproject/cortex/issues/2478)

I wanted to use this discussion to figure out what the best course of attack would be for tackling #2262. Since this will be a breaking change I'm wondering how we could keep compatibility for two versions before deprecating the old config. For the GCS client this is trivial since the configs are very similar. For S3 there are some features in the chunk client that do not exist in the TSDB objstore client. How would we go about deprecating those features without having a breaking config change for two versions?

Some ideas I have include:
- Separate config block for new clients, remove old blocks after two versions
- Compatibility config structs with fields for both, use the correct client based on the provided implementation. Remove deprecated fields until only the TSDB client config struct remains

**Comment by pstibrany** [comment]
> Separate config block for new clients, remove old blocks after two versions

I vote for this option. Seems like cleanest to me, at the cost of having two different object-store clients for two releases.

Note that Thanos objstore client currently doesn't support recursive listing used by our own clients, so that is something we may want to implement there first.

**Reply by jtlisi** [reply]
> Note that Thanos objstore client currently doesn't support recursive listing used by our own clients, so that is something we may want to implement there first.

Makes sense, I was able to make things work here: https://github.com/cortexproject/cortex/pull/2489. However, there have been some recent changes to the rules backend and I'm not sure if that PR is compatible anymore. In general though, it involves nesting lists which is not ideal. A better approach is needed

**Reply by pracucci** [reply]
> Separate config block for new clients, remove old blocks after two versions

I also vote for this.
